pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	resourceType     = "resource_type"
+	nodeName         = "name"
+	nodeProvisioner  = "provisioner"
+	nodeZone         = "zone"
+	nodeArchitecture = "arch"
+	nodeCapacityType = "capacity_type"
+	nodeInstanceType = "instance_type"
+	nodePhase        = "phase"
+	nodeLabels       = "node_labels"
+)
+
+var (
+	allocatableGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "allocatable",
+			Help:      "Node allocatable",
+		},
+		getLabelNames(),
+	)
+	requestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_requests",
+			Help:      "Node total requests",
+		},
+		getLabelNames(),
+	)
+	limitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_limits",
+			Help:      "Node total limits",
+		},
+		getLabelNames(),
+	)
+	overheadGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_overhead",
+			Help:      "Node total overhead",
+		},
+		getLabelNames(),
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(allocatableGaugeVec)
+	crmetrics.Registry.MustRegister(requestsGaugeVec)
+	crmetrics.Registry.MustRegister(limitsGaugeVec)
+	crmetrics.Registry.MustRegister(overheadGaugeVec)
+}
+
+func getLabelNames() []string {
```

Comment:
For static accessors, I often avoid the "get" and just name like "labelNames()". To me, get implies that you have to do something (e.g. network call, cache lookup) to get the result. This is very much a go-ism rather than a java-ism, at least as far as I've seen.

Reply:
Good point. I agree. Done